test/integration/model/cte.test.js

```diff
@@ -0,0 +1,812 @@
+          return User.create({ username: 'user1' }).then(function (user1) {
+            return User.create({ username: 'user1.1' }).then(function (user2) {
+              return User.create({ username: 'user1.2' }).then(function (user3) {
+                return User.create({ username: 'user1.1.1' }).then(function (user4) {
+                  return User.create({ username: 'user1.1.2' }).then(function (user5) {
+                    return User.create({ username: 'user1.2.1' }).then(function (user6) {
+                      return user1.setReports([user2, user3]).then(function () {
+                        return user2.setReports([user4, user5]).then(function () {
+                          return user3.setReports([user6]).then(function () {
+                            return User.findAll({
+                              cte: [{
+                                name: 'a',
+                                model: User,
+                                cteAttributes: ['level'],
+                                initial: {
+                                  level: 1,
+                                  where: { username: 'user1' }
+                                },
+                                recursive: {
+                                  level: { $add: [{ $cte: 'level' }, 1] },
+                                  next: 'reports',
+                                  order: [['level', 'DESC']],
+                                  where: {
+                                    cte: {
+                                      level: {
+                                        $lt: 2
+                                      }
+                                    }
+                                  }
+                                },
+                              }],
+                              cteSelect: 'a'
+                            }).then(function (selectedUsers) {
+                              expect(selectedUsers).to.have.length(3);
+                              expect(selectedUsers).to.contain.a.thing.with.property('username', 'user1');
+                              expect(selectedUsers).to.contain.a.thing.with.property('username', 'user1.1');
+                              expect(selectedUsers).to.contain.a.thing.with.property('username', 'user1.2');
+                            });
+                          });
+                        });
+                      });
+                    });
+                  });
+                });
+              });
+            });
+          });
+        });
+      });
+
+
+      it('can use a counter based on a model attribute in a recursive CTE', function () {
+        var User = this.sequelize.define('UserXYZ', {
+          username: DataTypes.STRING,
+          amount: DataTypes.INTEGER
+        });
+
+        User.hasOne(User, { as: 'report' });
+
+        return this.sequelize.sync({ force: true }).then(function () {
+          return Promise.join(
```

Comment:
I think it would be cleaner to just code the reporterId directly into the create statements

```js
User.hasOne(User, { as: 'report', foreignKey: 'reporterId' });

User.bulkCreate([
  { username: 'user1', amount: 12 },
  { username: 'user2', amount: 5, reporterId: 1 },
  { username: 'user3', amount: 2, reporterId: 2 },
  { username: 'user4', amount: 23, reporterId: 3 },
  { username: 'user5', amount: 13, reporterId: 4 },
  { username: 'user6', amount: 7, reporterId: 5 },
])
```